packages/pds/src/db/message-queue/index.ts

```diff
@@ -0,0 +1,288 @@
+import { RepoStructure } from '@atproto/repo'
+import Database from '..'
+import {
+  AddMember,
+  AddUpvote,
+  CreateNotification,
+  DeleteNotifications,
+  Message,
+  RemoveMember,
+  RemoveUpvote,
+} from './messages'
+import { sql } from 'kysely'
+import { dbLogger as log } from '../../logger'
+import { AuthStore } from '@atproto/auth'
+import * as repoUtil from '../../util/repo'
+import * as schema from '../../lexicon/schemas'
+import { TID } from '@atproto/common'
+import { MessageQueue } from '../types'
+
+type GetAuthStoreFn = (did: string) => AuthStore
+
+export class SqlMessageQueue implements MessageQueue {
+  private cursorExists = false
+
+  constructor(
+    private name: string,
+    private db: Database,
+    private getAuthStore: GetAuthStoreFn,
+  ) {}
+
+  async send(tx: Database, messages: Message | Message[]): Promise<void> {
+    const msgArray = Array.isArray(messages) ? messages : [messages]
+    if (msgArray.length === 0) return
+    const now = new Date().toISOString()
+    const values = msgArray.map((msg) => ({
+      message: JSON.stringify(msg),
+      createdAt: now,
+    }))
+
+    await tx.db.insertInto('message_queue').values(values).execute()
+
+    this.process()
+  }
+
+  private async ensureCursor(): Promise<void> {
+    if (this.cursorExists) return
+    await this.db.db
+      .insertInto('message_queue_cursor')
+      .values({ consumer: this.name, cursor: 1 })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    this.cursorExists = true
+  }
+
+  async process(): Promise<void> {
+    await this.ensureCursor()
+    const keepGoing = await this.db.transaction(
+      async (dbTxn): Promise<boolean> => {
+        let builder = dbTxn.db
+          .selectFrom('message_queue_cursor as cursor')
+          .innerJoin('message_queue', 'message_queue.id', 'cursor.cursor')
+          .selectAll()
+        if (this.db.dialect !== 'sqlite') {
+          builder = builder.forUpdate()
+        }
```

Comment:
You could add a `skipLocked()` here so that if there's contention for the cursor, one consumer gets it and the other just bails rather than wait for it to open up.

Reply:
i think we actually want to wait for it to open up. we basically ping the consumer on every event push to try & process exactly 1 event. So it should wait until the cursor is freed up & then give it a go

Reply:
Ahh, okay got it.  How does `keepGoing` play into that?

Reply:
ah not sure if you saw, but `keepGoing` actually got ripped out.

Yeah beforehand, i was throwing way to many handlers at this & they'd all just hang out at the lock with their hands open until the DB got back to them & said "sorry buddy no cursor for you"

Reply:
Ahhh right on!